internal/wasm/host.go

```diff
@@ -4,90 +4,112 @@ import (
+	if moduleName != "" {
+		m = &Module{NameSection: &NameSection{ModuleName: moduleName}}
+	} else {
+		m = &Module{}
+	}
+
+	funcCount := uint32(len(nameToGoFunc))
+	memoryCount := uint32(len(nameToMemory))
+	exportCount := funcCount + memoryCount
+	if exportCount > 0 {
+		m.ExportSection = make(map[string]*Export, exportCount)
+	}
+
+	if funcCount > 0 {
+		if err = addFuncs(m, nameToGoFunc); err != nil {
+			return
+		}
+	}
+
+	if memoryCount > 0 {
+		if err = addMemory(m, nameToMemory); err != nil {
+			return
 		}
 	}
+	return
+}
 
-	m := &Module{
-		NameSection:         &NameSection{ModuleName: moduleName, FunctionNames: make([]*NameAssoc, 0, hostFunctionCount)},
-		HostFunctionSection: make([]*reflect.Value, 0, hostFunctionCount),
-		ExportSection:       make(map[string]*Export, hostFunctionCount),
+func addFuncs(m *Module, nameToGoFunc map[string]interface{}) error {
+	funcCount := uint32(len(nameToGoFunc))
+	funcNames := make([]string, 0, funcCount)
+	if m.NameSection == nil {
+		m.NameSection = &NameSection{}
 	}
+	m.NameSection.FunctionNames = make([]*NameAssoc, 0, funcCount)
+	m.FunctionSection = make([]Index, 0, funcCount)
+	m.HostFunctionSection = make([]*reflect.Value, 0, funcCount)
 
-	// Ensure insertion order is consistent
-	names := make([]string, 0, hostFunctionCount)
+	// Sort names for consistent iteration
 	for k := range nameToGoFunc {
-		names = append(names, k)
+		funcNames = append(funcNames, k)
 	}
-	sort.Strings(names)
+	sort.Strings(funcNames)
 
-	for idx := Index(0); idx < hostFunctionCount; idx++ {
-		name := names[idx]
+	for idx := Index(0); idx < funcCount; idx++ {
+		name := funcNames[idx]
 		fn := reflect.ValueOf(nameToGoFunc[name])
 		_, functionType, _, err := getFunctionType(&fn, false)
 		if err != nil {
-			return nil, fmt.Errorf("func[%s] %w", name, err)
+			return fmt.Errorf("func[%s] %w", name, err)
 		}
 
 		m.FunctionSection = append(m.FunctionSection, m.maybeAddType(functionType))
 		m.HostFunctionSection = append(m.HostFunctionSection, &fn)
 		m.ExportSection[name] = &Export{Type: ExternTypeFunc, Name: name, Index: idx}
 		m.NameSection.FunctionNames = append(m.NameSection.FunctionNames, &NameAssoc{Index: idx, Name: name})
 	}
-	return m, nil
+	return nil
 }
 
-func (m *Module) maybeAddType(ft *FunctionType) Index {
-	for i, t := range m.TypeSection {
-		if t.EqualsSignature(ft.Params, ft.Results) {
-			return Index(i)
+func addMemory(m *Module, nameToMemory map[string]*Memory) error {
+	memoryCount := uint32(len(nameToMemory))
+
+	// Only one memory can be defined or imported
+	if memoryCount > 1 {
+		memoryNames := make([]string, 0, memoryCount)
+		for k := range nameToMemory {
+			memoryNames = append(memoryNames, k)
 		}
+		return fmt.Errorf("only one memory is allowed, but configured [%s]", strings.Join(memoryNames, ","))
 	}
 
-	result := m.SectionElementCount(SectionIDType)
-	m.TypeSection = append(m.TypeSection, ft)
-	return result
-}
-
-func (m *Module) validateHostFunctions() error {
```

Comment:
validateHostFunctions() was unneeded as they are only instantiated here and validated on insertion